### DISCUSSION TITLE
Energy Usage graph colour select or customisation option for stack

### DISCUSSION THREAD

**Discussion by groove200** [home-assistant/frontend#12397] [Accessibility] [answered]
I've seen it asked on the forums, but doesn't look like the request ever made it here. It would be so much better if we had more control over the energy dashboard, especially the usage graph. I've nothing against blue, but it can get difficult to read if we want to add more trackers to this. (I know the individual is split out with different colours, but it is nice to be able to see in the stack the biggest contributor etc on a time frame) plus, colour blind people are going to have a really hard time distinguishing shades like this.

Feels strangely 'locked down' to the rest of the ecosystem.

https://community.home-assistant.io/t/energy-tab-more-colors-for-graphs/336089

**Comment by jmcollin78** [comment]
I have a small variation on colors but I would rather prefer to have some friendly color for this dashboard on energy source:
![Capture d’écran 2022-12-11 à 18 40 53](https://user-images.githubusercontent.com/1712374/206919615-791615d2-9014-4c2b-98fa-554c2c027a77.png)

Ideally colors should be defined at dashboard definition level :
![Capture d’écran 2022-12-11 à 18 44 04](https://user-images.githubusercontent.com/1712374/206919741-1bdcdc85-44a9-4893-84af-1c6fcfde7c8f.png)

**Reply by madmalkav** [reply]
Same problem here, not very easy to look at for my poor, tired eyes 😅

**Reply by dennis3484** [reply]
Not very easy to look at even when not tired 🥹

**Reply by lecointrem** [reply]
même problème, avec l'abonnement tempo d'EDF
![image](https://user-images.githubusercontent.com/54736243/214139275-044c91b8-ed86-4d36-925b-5d6d5f3acdc9.png)

**Comment by bjornarfjelldal** [comment]
+1 on this! I have 25+ sources (monitoring each individual breaker) that fit better in this kind of stacked bar graph per hour instead of the "individual devices" list below it, because it gives richer hour-to-hour insights on consumption and also cost. Or a 2.0 version of energy monitoring and analysis - the Energy tab has a lot of potential. Thanks for your work :)
<img width="987" alt="image" src="https://user-images.githubusercontent.com/52956570/211028890-c98966d5-79f6-416a-b29b-286657466fae.png">

**Comment by mathieucarbou** [comment]
I would like some control too.. In France, there are several tarifs and here is my current display:
![image](https://user-images.githubusercontent.com/61346/217260373-270f5475-8dcd-423b-9f02-e9e5f8fe7c92.png)
It is then very hard to distinguish in the bar chart

**Comment by me1337me** [comment]
+1 for a color picker

**Comment by mathieucarbou** [comment]
I've created an issue regarding the problem caused by the current implementation here: https://github.com/home-assistant/frontend/issues/15384
This feature request would solving the issue.

**Comment by groove200** [comment]
Nearly a year and the energy features still get no love and yet it's one of the best features (or could be) for me at least.

Every month I read the release notes in anticipation of some enhancements, every month im dissapointed. Such a shame as it really feels like a neglated part of the system that gets overlooked in favour of the latest automation buzz topics

**Comment by GizmoGoody** [comment]
+1 The closeness of the colors and the lack of filter on the data popup specific to that consumer's section of the column makes it nearly impossible to discern what belongs to which consumer.

![image](https://user-images.githubusercontent.com/46235745/219138469-3fc6d37c-d6b5-4b79-8aae-de5701a58fc4.png)
![image](https://user-images.githubusercontent.com/46235745/219138654-0054727e-0e60-49b5-8755-ecfd3bcf744b.png)
![image](https://user-images.githubusercontent.com/46235745/219139291-2a706fcf-ac4b-4ef2-be86-98adc01a7fae.png)

**Comment by mikke49** [comment]
+1

**Comment by saltamura-tic** [comment]
Same problem. Thanks

**Comment by pbranly** [comment]
Same problem thanks

**Comment by Binifada1956** [comment]
Also here theme Problem. It would be really great to change the Color individuell!

![image](https://user-images.githubusercontent.com/99736262/221365533-f65e911d-475e-416a-811e-54c89b40948b.png)

**Comment by StrSebastian** [comment]
same here

**Comment by digimp** [comment]
+1 for a color picker

**Comment by Loic691** [comment]
+1 for change color

**Reply by pbranly** [reply]
Same +1

**Reply by PieTr031** [reply]
+1 for change color 2 :

**Comment by ALaDoffe** [comment]
Same +1

**Comment by roumano** [comment]
As it's thread have been opened as discussions not as a bug, i think this issue is not look & take care by dev.
   A issue should be open to have a chance to cover this issue

**Reply by tiramiseb** [reply]
@roumano it already exists: https://github.com/home-assistant/frontend/issues/15384

Moreover, when we open a feature request, it creates a discussion. Do you imply that developers do not take care of feature requests.

**Comment by Ba-pt0u** [comment]
+1 to this request

**Comment by Mika255** [comment]
+1

**Comment by alimnette** [comment]
+1

**Comment by drzoidberg33** [comment]
This would be great!

The newly added card for self-sufficiency is a particularly bad one:

![image](https://github.com/home-assistant/frontend/assets/1313205/cf4eadd8-2e46-4c7a-839f-0f5de5f0a423)

One side is blue and the other side is purple. Looks like two different blues to me 🤣

**Reply by karwosts** [reply]
That is the grid-neutrality card, and it is already themable today, just add the following to your theme with your favorite RGB colors:

```
      energy-grid-consumption-color: "#ff0000"
      energy-grid-return-color: "#0000ff"
```

![image](https://github.com/home-assistant/frontend/assets/32912880/9f368c13-3a45-4b99-b8d8-5e549380c069)

**Comment by groove200** [comment]
Think we are all wasting our time here. Energy and accessibility options are just not cool enough these days.

**Comment by mdaitn** [comment]
+1

**Comment by Nitrouze** [comment]
+1

**Comment by ahomberger** [comment]
+1 to this request

**Comment by john8329** [comment]
Some customization would fit nicely there in my opinion, Solaredge rightly puts the "imported energy" as red which means "bad thing happening" intuitively, it's easier to notice than blue. My only option at the moment is to build a custom graph for it to match the SE dashboard as closely as possible.

**Reply by karwosts** [reply]
You can already set the base colors of all the energy elements with a theme, these are the variables that can be used:

```
energy-battery-in-color
energy-battery-out-color
energy-gas-color
energy-grid-consumption-color
energy-grid-return-color
energy-non-fossil-color
energy-solar-color
energy-water-color
```

It is only the individual gradient elements which are not currently customizable. But if you just want to swap red to blue or something like that it is currently doable.

**Comment by gianlucaf81** [comment]
absolutely essential feature! Enjoy the holidays but then implement it now. Thank you!

**Reply by rozanders** [reply]
Yes, the format of the dashboard is very static.
Colors or pre defind cards can't be changed.
Or You have to program it from start.
Pls, make it a more flexibele.

**Comment by Jeancstl** [comment]
+1 !

**Comment by S474N** [comment]
Need also hot/cold watter, please!

**Comment by karwosts** [comment]
#17527 Just got merged for the 2023.9 beta, it will allow applying theme colors for each individual sub-item in the energy usage chart stack. 

I didn't get a chance to do this feature for all the other energy cards yet, but I will try to get that in for next month. 

If anyone is going to be trying out the beta today, please give it a shot and let me know how it works!

**Reply by drzoidberg33** [reply]
Thanks for thinking of us colour-challanged folk 😆

**Reply by pics195** [reply]
Merci beaucoup pour la prise en compte. A nous les couleurs !

**Reply by GizmoGoody** [reply]
Did this make the 2023.9 release? Example configuration docs?

**Reply by karwosts** [reply]
Only the `energy-usage-graph` card got the update in 2023.9, the rest of the dashboard cards (gas/solar/water/sources-table) is waiting for #17826, probably in 2023.10. 

For configuring energy-usage-graph, you can add the following values to your theme:

```
energy-grid-return-color-#
energy-grid-consumption-color-#
energy-battery-in-color-#
energy-battery-out-color-#
```

Where # is the number of the 0-based index you want to color. 

e.g.
```
      energy-grid-consumption-color-0: "#00ff00"
      energy-grid-consumption-color-1: "#ff00ff"
      energy-grid-consumption-color-2: "#00ffff"
      energy-grid-consumption-color-3: "#ffff00"
      energy-grid-return-color-0: "#31BD23"
      energy-grid-return-color-1: "#0ff8ec"
      energy-battery-in-color-0: "#0000ff"
      energy-battery-in-color-1: "#00ffff"
```

General themes info at: https://www.home-assistant.io/integrations/frontend/#defining-themes

**Reply by karwosts** [reply]
Why do you have to "redefine a complete theme"? You can just add these lines to an existing theme. 

And no there's no other option. The way to customize UI element colors in HA is through theme variables. Don't quite understand what the concern is here.

**Reply by mathieucarbou** [reply]
I think I got it working but I had to update the specific global theme I use.

Example for EDF Tempo in France:

```
  energy-grid-consumption-color-0: blue
  energy-grid-consumption-color-1: darkblue
  energy-grid-consumption-color-2: white
  energy-grid-consumption-color-3: lightgrey
  energy-grid-consumption-color-4: red
  energy-grid-consumption-color-5: darkred
```

![image](https://github.com/home-assistant/frontend/assets/61346/8c200ea1-da60-431e-b84c-8d4f6ec248b1)

But I still have the old colors in the sources:

![image](https://github.com/home-assistant/frontend/assets/61346/e8d5635d-81be-4d7b-8ba8-97f8ade088ac)

**Reply by karwosts** [reply]
> But I still have the old colors in the sources:

Yes like I mentioned that is the energy-sources-table, and the update for that will come after the referenced PR gets merged. Didn't make 2023.9

**Reply by karwosts** [reply]
Also I would maybe recommend using hex colors instead of color names. If you use names it sort of looks like it works but the bar borders, tooltips, and comparison bars don't color correctly, since they do some raw manipulation by adding alpha values to the hex codes from the theme.

**Reply by pics195** [reply]
It works, but like mathieucarbou, no modification for the source. Damage!

**Reply by Binifada1956** [reply]
![image](https://github.com/home-assistant/frontend/assets/99736262/99514a1e-fff8-45e4-955b-2e28ac49c96f)

Hi, I think it works. Now only the colours of the sources have to be synchronised with the energy users.

Thank you "karwosts" for your work!

**Reply by pbranly** [reply]
Hi
This is working fine for grid_consumption . Thanks
I tried to use colors as well for water and gas but it seems that colors don’t follow 
Are they supported ?
Phil

**Reply by S474N** [reply]
Also waiting for water (I need to differentiate between hot and cold water by color).

**Reply by karwosts** [reply]
Water and gas colors is coming in the next PR (#17826)

**Reply by titom43** [reply]
Not sure to understand. How do you apply themes? 

EDIT: I mean how do you choose the color for each power source? (what IDs to mention??)

**Reply by karwosts** [reply, marked as answer]
Looks like we're fully merged for the rest of the cards now for 2023.11. 

Full list of variables added that can be set (replace 0 with the index number you want to modify):

--energy-grid-return-color-0
--energy-grid-consumption-color-0
--energy-battery-in-color-0
--energy-battery-out-color-0
--energy-solar-color-0
--energy-gas-color-0
--energy-water-color-0

**Reply by Loic691** [reply]
Hi great !
I have an issue with the old consumption of last year. The data are display in black color...
Is there a special config for this ?

Here's my cuctom code
```
frontend:
  themes:
    Custom Home Assistant:
      energy-grid-consumption-color-0: 'lightblue'  # hcjb bleu
      energy-grid-consumption-color-1: 'DodgerBlue' # hpjb bleu
      energy-grid-consumption-color-2: 'lightgrey'  # hcjw blanc
      energy-grid-consumption-color-3: 'grey'       # hpjw blanc
      energy-grid-consumption-color-4: 'lightcoral' # hcjr rouge
      energy-grid-consumption-color-5: 'Crimson'    # hpjr rouge
```
![image](https://github.com/home-assistant/frontend/assets/60837526/601a54be-129a-4360-a9d0-70728e38b3e2)

**Reply by pvdeye** [reply]
please tell me the variable from individual?

**Comment by lecointrem** [comment]
On a tous le même problème Mathieu...

Le mar. 7 févr. 2023 à 14:39, Mathieu Carbou ***@***.***> a
écrit :

**Comment by smarthomecircle** [comment]
> Not sure to understand. How do you apply themes?
> 
> EDIT: I mean how do you choose the color for each power source? (what IDs to mention??)

@titom43 you can check my video here: https://www.youtube.com/watch?v=9c2XUjobVB8. I have shown how to do it for themes as well. Hope it helps

**Reply by titom43** [reply]
@smarthomecircle Thank you so much for your video. It was very clear. You saved my time mentioning that the colors are not yet applied to the legend...

**Comment by roumano** [comment]
> energy-grid-consumption-color-5
as you have 7 source input, maybe need to add a `energy-grid-consumption-color-6`, can you try  ?

**Reply by Loic691** [reply]
I think I done an error...
I have now 8 sources but the two first are old sources...
I delete the two first from the energy panel config.
Now I have just 6 sources but the index color seams not good for data from now...
I delete the sensor data for my old sources...

![image](https://github.com/home-assistant/frontend/assets/60837526/cd06e584-6b6f-4273-887c-c4e475df50f5)

**Reply by Loic691** [reply]
The color are now correct. Ouf... Maybe BDD auto synchronzation ? EDIT => No it's a problem of cache browser... Cool
When I select data from last here, it's ok too.
**The black color appears just when I enter in comparative mode...**
I think it's a bug or miss config ? No link with my deletion of 2 first input in energy panel
![image](https://github.com/home-assistant/frontend/assets/60837526/fcf19e76-c7f0-4377-b7a8-fea981962375)

**Reply by Loic691** [reply]
For completing information. I have no problem with the default theme of HA in comparative mode.
I think it's a bug and we have to adjust other colors for comparative mode...

**Reply by karwosts** [reply]
The issue is compare is expecting colors to be in the hex format #rrggbb, not a string name.

It gets the compare color by appending an alpha channel to the rrggbb color so it becomes `#rrggbb7f`, it's half transparent. 

So using a name like "red" will not work, while #ff0000 will work correctly.

**Reply by Loic691** [reply]
It's ok .... Thank you very much !

**Reply by karwosts** [reply]
The black color bars should be fixed in a more recent version, you can use color names now instead of hex.

**Comment by ManuelKugelmann** [comment]
BTW: the devices take their colors from `--graph-color-1 --graph-color-2 ...`

**Reply by Chrismettal** [reply]
Thanks, wasn't able to find that info anywhere!

**Comment by pvdeye** [comment]
![Unbenannt](https://github.com/home-assistant/frontend/assets/124249318/774400fb-cdc5-4cad-8253-718e1e911fb8)
How i can change consumed battery color and consumed solar color?

**Reply by karwosts** [reply]
Believe the theme vars for consumed solar is `--energy-solar-color`, and for consumed battery is `--energy-battery-out-color`

**Reply by pvdeye** [reply]
no sorry energy-solar-color is only the color from inverter-solar-power but not the consumed solar power
The same on consumed battery out

**Reply by ManuelKugelmann** [reply]
Afaik the Energy Usage Graph does not show the full inverter solar power, but only the consumed part of it colored by energy-solar-color, the parts that are not consumed but put into the battery or grid are colored by energy-battery-in-color and energy-grid-return-color. Consumed grid energy is colored by energy-grid-consumption-color and consumed battery energy is energy-battery-out-color.

I have not yet tested how charging the battery with grid energy is displayed, but I would guess, that the grid energy is considered consumed.

All consumed energy is stacked above the 0-line, all energy that is either exported into the grid or put into the battery is below the 0-line. Thus the the total height of the stack represents the total energy throughput, the part above 0 consumption and the part below 0 is export and storage. Imho an nice way to put a lot of information into a single graph.

**Comment by jmcollin78** [comment]
Hello, does someone have the color customisation available on mobile device ?
It works fine for me in computer Chrome but not in the app Android nor IPad.

**Reply by pics195** [reply]
Bonjour 
Il faut sélectionner le même thème sur votre application Android ou Ipad que sur votre PC. Je parle du thème général.

**Reply by jmcollin78** [reply]
Ah merci ! Oui ca marche, je ne pensais pas que c'était lié au device.

**Comment by madmalkav** [comment]
O see this has been marked as complete. Does that lean that editing config files will be the only way to modify those colors? No plans to provide a way to do it in the UI in the future?

**Reply by silamon** [reply]
I've reopened. It should be added to the UI, but that wasn't clear in the initial post and the selected answer.

**Comment by kkevindev** [comment]
I think a feature where you can edit these in the front-end would be a must have addition for color blind people.

**Reply by Matthias84** [reply]
Something like Grafana does it would be great. But I guess this would require a deeper refactoring if you want to allow it for dashboards in general?
